bpf/bpf_overlay.c

```diff
@@ -323,6 +329,49 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 					      CTX_ACT_DROP, METRIC_INGRESS);
 	return ret;
 }
+
+#ifdef ENABLE_VTEP
+/*
+ * ARP responder for ARP requests from VTEP
+ * Respond to remote VTEP endpoint with cilium_vxlan MAC
+ */
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_ARP)
+int tail_handle_arp(struct __ctx_buff *ctx)
+{
+	union macaddr mac = NODE_MAC;
+	union macaddr smac;
+	__be32 sip;
+	__be32 tip;
+	__u32 monitor = 0;
+	int i;
+	int ret;
+	struct remote_endpoint_info *info;
+	struct bpf_tunnel_key key = {};
+
+	if (unlikely(ctx_get_tunnel_key(ctx, &key, sizeof(key), 0) < 0))
+		return DROP_NO_TUNNEL_KEY;
+
+	if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
+		return CTX_ACT_OK;
+	if (!__lookup_ip4_endpoint(tip))
+		return CTX_ACT_OK;
+	ret = arp_prepare_response(ctx, &mac, tip, &smac, sip);
+	if (unlikely(ret != 0))
+		return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);
+
+	info = lookup_ip4_remote_endpoint(sip);
+	if (!info)
+		return CTX_ACT_OK;
```

Comment:
Logically it makes more sense to do all of the error checks first (including looking up the remote endpoint where we need to send the traffic to), *then* modify the packet and transmit it back. This way, for instance if we trigger `send_drop_notify_error()`, we will send that notification with the original packet contents so it will look like "Cilium received the ARP request, could not process it, and decided to drop it" instead of "Cilium received an ARP response, could not process, and decided to drop it".

Reply:
@joestringer good point, I think will return `CTX_ACT_DROP` instead of `send_drop_notify_error()` after `arp_prepare_response` since the packet is already changed after that, no point to notify it with changed packet or pass it to stack.

Reply:
All of the drop cases in this new tail call should return the result of `send_drop_notify_error()`, passing the drop reason in as the third parameter. This way, the drop notification will be sent to userspace with the reason for the drop.

Reply:
ok, that sounds good, I am still not sure if the last `return CTX_ACT_DROP;` is really required.

```
        if (unlikely(ctx_get_tunnel_key(ctx, &key, sizeof(key), 0) < 0))
                return send_drop_notify_error(ctx, 0, DROP_NO_TUNNEL_KEY, CTX_ACT_DROP, METRIC_EGRESS);
        
        if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
                return CTX_ACT_OK;
        if (!__lookup_ip4_endpoint(tip))
                return CTX_ACT_OK;
        info = lookup_ip4_remote_endpoint(sip);
        if (!info)                                                      
                return CTX_ACT_OK;

        ret = arp_prepare_response(ctx, &mac, tip, &smac, sip);
        if (unlikely(ret != 0))
                return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);

        for (i = 0; i < VTEP_NUMS; i++) {
                if (info->tunnel_endpoint == VTEP_ENDPOINT[i])
                        return __encap_and_redirect_with_nodeid(ctx, info->tunnel_endpoint,
                                                                        info->sec_label, monitor);
        }

        return CTX_ACT_DROP;
```

Reply:
@joestringer if I remove the last return, errors 

```
bpf_overlay.c:382:1: error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]
```

so if I keep the last  `return CTX_ACT_DROP;`, should I also do `send_drop_notify_error` ? but for what drop reason?  `DROP_UNKNOWN_l3` ?

Reply:
Yeah, that should also `send_drop_notify_error()`. Looking through the codebase at the error message for `DROP_UNKNOWN_L3`, it looks like that's a good choice for the drop code here:

```
$ git grep -i unknown.*l3
api/v1/flow/flow.pb.go: DropReason_UNKNOWN_L3_TARGET_ADDRESS                             DropReason = 150
api/v1/flow/flow.pb.go:         150: "UNKNOWN_L3_TARGET_ADDRESS",
api/v1/flow/flow.pb.go:         "UNKNOWN_L3_TARGET_ADDRESS":                             150,
api/v1/flow/flow.proto:    UNKNOWN_L3_TARGET_ADDRESS = 150;
api/v1/observer/observer.pb.go:const DropReason_UNKNOWN_L3_TARGET_ADDRESS = flow.DropReason_UNKNOWN_L3_TARGET_ADDRESS
bpf/bpf_host.c:         ret = send_drop_notify_error(ctx, identity, DROP_UNKNOWN_L3,
bpf/bpf_host.c:         ret = DROP_UNKNOWN_L3;
bpf/bpf_host.c:         ret = DROP_UNKNOWN_L3;
bpf/bpf_host.c:         ret = DROP_UNKNOWN_L3;
bpf/bpf_host.c:         ret = DROP_UNKNOWN_L3;
bpf/bpf_lxc.c:          ret = DROP_UNKNOWN_L3;
bpf/bpf_lxc.c:          ret = DROP_UNKNOWN_L3;
bpf/bpf_lxc.c:          ret = DROP_UNKNOWN_L3;
bpf/bpf_overlay.c:              ret = DROP_UNKNOWN_L3;
bpf/bpf_overlay.c:              ret = DROP_UNKNOWN_L3;
bpf/lib/common.h:#define DROP_UNKNOWN_L3                -139
bpf/lib/proxy.h:        ret = DROP_UNKNOWN_L3;
pkg/monitor/api/drop.go:        150: "Unknown L3 target address",
```

Reply:
ok, I pushed the changes.